Update dataAddr whenever GC moves indexable object

bragaigor opened the pull request:
Update GC code to make sure whenever an indexable object moves, the object will always contain the correct `dataAddr` value, which is the address pointing right after array header. There's no need to update array copy or array clone in GC side since we either initialize header before copying or we take JIT path.

We introduce pre/postObjectCompact in EnvironmentDelegate as well as struct GCmovedObjectHashCode which is being moved from OMR_VMThread to EnvironmentDelegate.

This is phase 2 of https://github.com/eclipse/openj9/issues/11438

Signed-off-by: Igor Braga <higorb1@gmail.com>

bragaigor commented:
cc: @amicic @dmitripivkine

amicic commented:
This should actually go _before_  OMR changes, since this one introduces setObjectDataAddr that OMR uses (OMR PR will be ok, but eventually OpenJ9's OMR acceptance will fail).

I don't see dependency on OMR (what is a more typical case)

bragaigor commented:
That makes sense, we'll be hitting the unimplemented `setObjectDataAddr` if this doesn't go first. I'll update the comments

bragaigor commented:
Optimally, these 2 PRs should be merged as close to each other as possible. If this one goes first, some platforms will complain of unimplemented `setObjectDataAddr` in ObjectModel (tested that before). And if OMR change is merged first then we'll hit the `Assert_MM_unreachable()` from `setObjectDataAddr`

bragaigor commented:
Edit: The issue I was having was related to `CompactScheme` and not having `setObjectDataAddr` implemented in OMR side for ObjectModel. But since that will be introduced with OMR PR, we do not have a cyclic reference

amicic commented on runtime/gc_glue_java/ArrayletObjectModel.hpp:
```diff
@@ -923,6 +923,37 @@ class GC_ArrayletObjectModel : public GC_ArrayletObjectModelBase
+		bool isDiscontiguous = (layout != InlineContiguous);
+
+		if (isDiscontiguous) {
+			setDataAddrForDiscontiguous(j9ArrayPtr, address);
```
hm, why do we need this branch (and then why passing address)!?

bragaigor replied:
For gencon there can still exist zero sized arrays. The address is there in case of double mapping and the future off-heap technology. But, whenever that's the case we'll know that the indexable object we're dealing with is indeed discontiguous, which we could call `setDataAddrForDiscontiguous` instead. With that being said we don't really need to pass `address` here. But we still need the branch here because we don't know the layout of the indexable object at this point. We only know that it's indexable, or am I missing something?

amicic replied:
I'm thinking to specialize this method and do just update/fixup (and rename it accordingly0 . That would be required only for contiguous (or perhaps for 0-sized too - does 0-sized array have dataPtr point to the end of header or it's NULL?).

As a more generic set method, yes, this would be ok.

bragaigor replied:
I see, maybe we could have something like `updateObject` ? And that's the thing, the call sites of `setObjectDataAddr` which are CopyForward and WriteOnceCompactor, they are not aware if the object is going to be contiguous and discontiguous. If we only want to update contiguous objects here we would have to make the distinction between contiguous and disgontiguous as well. So I would say to updated zero sized as well, that way all `dataAddr` indexable object headers will be consistent, does this make sense?

amicic commented on runtime/gc_vlhgc/WriteOnceCompactor.cpp:
```diff
@@ -743,6 +743,14 @@ MM_WriteOnceCompactor::evacuatePage(MM_EnvironmentVLHGC *env, void *page, J9MM_F
 				}
 				if(_extensions->objectModel.isIndexable(newLocation)) {
 					updateInternalLeafPointersAfterCopy((J9IndexableObject*) newLocation, (J9IndexableObject*) objectPtr);
+#if defined(J9VM_ENV_DATA64)
```
Let's have this opportunity to make code a bit more readable and more consistent (with OMR compactor), and try introduce WOC::preObjectMove/postObjectMove  methods (move everything about remembering, hashing etc. in them)

You may want to rely on objectMode.preMove/postMove, but be completely sure they do the same what we do here. They will be marginally slower (since accessing thread struct, rather than stack), but this is anyhow relatively infrequent/slow path, so I'm fine with it.

If not possible then, you'll have to extend signatures of WOC::preObjectMove/postObjectMove to pass out/in hash/move flags.

And now to be really annoying, I'll say that I don't like that movedObjectHashCode struct is in OMR Thread , while all other hash/move knowledge is very OpenJ9 specific. The struct should be in OpenJ9's Thread (what would require that objectMode.preMove/postMove deal with Environement and OMR thread, and they extract OpenJ9's Thread from Env). What do you think? Even if we want to change that, perhaps it would be cleaner to do separately!?




unlike in OMR (which allows different language to have different actions/flags for hashing & movement), you don't have to rely on objectModel.preMove/postMove

bragaigor replied:
I see what you mean, from a quick look it seems that it would be possible to substitute those calls with objectModel.preMove/postMove . I haven't found any usages of `movedObjectHashCodeCache` in OMR side either so moving it into J9VMThread does make sense. If we introduce `struct movedObjectHashCode movedObjectHashCodeCache;` in J9VMThread in this PR, we could just remove the equivalent one in OMR with the OMR PR.

bragaigor replied:
I have good and bad news. objectModel.preMove/postMove do exactly as what we want here hence we can substitute the code before and after `memmove` with those. I've also created `preObjectMove` and `postObjectMove` to be consistent with OMR compactor. The bad news is that we can substitute `OMR_VMThread` with `J9VMThread` and transfer `movedObjectHashCode` over; however, that would require that we update the objectModel.preMove/postMove API to accept `J9VMThread` instead of `OMR_VMThread`. With that, OMR compactor objectModel.preMove/postMove API complains that it must also use `J9VMThread`, which would lead us to also updating OMR compactor to use `J9VMThread` which is something we don't want? 

I tried to get access to `J9VMThread` through `OMR_VMThread` but I don't think that's possible.

amicic replied:
Right, but perhaps this is still doable..... 

movedObjectHashCode is used only within GC code, so it's even better to be inside GC (and J9) specific struct MM_EnvironmentDelegate (or GC_Environment? - honestly, don't know why they are separate). And also declared somewhere in GC specific file. 

objectModel.preMove/postMove would then stay with OMR specific parameter (OMR_VMThread or EnvironmentBase), from where we know how to access language specific delegates.

bragaigor replied:
I see, I think we can get access to those via `OMR_VMThread`, looking into it

bragaigor replied:
We can actually access `J9VMThread` from `OMR_VMThread` via member `_language_vmthread` (should have looked harder before). So I'll put it under `J9VMThread` and since we cannot name it the same and this is J9/GC specific I was thinking about `J9movedObjectHashCode` or maybe `GCmovedObjectHashCode`?

amicic replied:
I still prefer it in EnvDelegate. It's only used in GC code.

amicic replied:
then even change objectModel.preMove/postMove to pass EnvBase (again same argument, we don't need anything from OMRVMThread, everything is in GC structs).

For that, to break mutual dependancy with OMR you'll need to clone objectModel.preMove/postMove, and remove the old ones with a followup J9 PR.

bragaigor replied:
Now in order to get access to EnvDelegate we need to go through EnvBase, but EnvDelegate member in EnvBase is protected and defined in OMR side (it seems this is the only way we can access EnvDelegate). In that case we would need to update OMR side first before we can access `movedObjectHashCode` inside EnvDelegate, unless I'm missing something?

amicic replied:
you can put the struct in GC_Environment component of EnvDelegate, that seems to have a public getter in EnvBase:

	MMINLINE GC_Environment *getGCEnvironment() { return _delegate.getGCEnvironment(); }

bragaigor replied:
Everything is good with this approach except that I started having issues when trying to inline the new objectModel.preMove/postMove . Now that EnvBase is one of the parameters, objectModel.preMove/postMove bodies must go in the source cpp file instead of the header file. I need to do that because I can't include EnvBase from the header file; I get several `invalid use of incomplete type` error which are probably due to cyclic imports. I've had this issue in the past (trying to include `EnvironmentBase.hpp` in `ObjectModel.hpp`). And the inline error I get for not having the body in header file is:

```
error: inlining failed in call to always_inline 'void GC_ObjectModel::preMove...': function body not available
```
What would you recommend going forward? Should I deal with these cyclic reference issue?

amicic replied:
We are finally realizing why movedObjectHashCode struct (that is language specific) put into an OMR thread local struct, to start with. Without trying to understand everything, there are cycle references between EnvBase (what includes ExtBase) and ObjectModel (that is embedded in ExtBase). OMR thread local struct is out of that loop and also can be obtained by caller before we enter the loop by calling objectModel.preMove/postMove....

We might give up to this part of changes, afterall. But there might be some hope if we take preMove/postMove from ObjecdtModel into EnvBase  (even if that works, it's generally against the trend where we try to leave EnvBase lean)

bragaigor replied:
I think I got something going, followed by your advice. I’ve included preMove/postMove inlined implementation in `EnvironmentDelegate`, and have the according wrappers at `EnvironmentBase`. Now `struct movedObjectHashCode` resides in `GC_Environment`. Everything works but in order to this correctly (without having mutual dependency with OMR) I think we would need a 4 way set of PRs since we need the wrappers in place in `EnvironmentBase` which is in OMR side:
1. This PR first, where we introduce preMove/postMove inlined in `EnvironmentDelegate`
2. OMR PR https://github.com/eclipse/omr/pull/5795 where we introduce wrappers at `EnvironmentBase`. At this point we can update OMR compactor to use the new preMove/postMove from `EnvironmentDelegate`.
3. Another OpenJ9 PR that will update `WriteOnceCompactor` preMove/postMove to use the new `EnvironmentBase` preMove/postMove. And at this point we can remove ObjectModel.preMove/postMove. 
4. Last OMR PR where we remove the obsolete `struct movedObjectHashCode`. We can probably make this a beginner’s issue?

How does this sound?

amicic replied:
Yes, this is this looks good. 

But then, while we definitely did cleaner separation between OMR/OpenJ9, we did violate a tacit rule that Env/Ext should be lean. I'll leave it up to you, what you prefer more (or find a better solution).

If we stick with this, I think it would be ok to cache ExtBase in EnvDelegate (it already caches EnvBase, from which you can get ExtBase, but caching it explicitly is even better).

bragaigor replied:
Sounds good, let me give another thought to this and see if this is indeed better than having pre/postMove in ObjectModel

bragaigor replied:
I found another possible place to put preMove/postMove which is CompactSchemeFixupObject.hpp but it would be out of scope and no straight forward way to access unless we make them static methods. Given our circumstances/constraints and keeping the separation between OMR/OpenJ9, placing preMove/postMove in `EnvironmentDelegate` seems to be the best approach. It feels like preMove/postMove will be a little bit out of place (inside `EnvironmentDelegate`) but I don't see any other reasonable location to place it. The other 2 options would be:  
- leave things as is, which is to keep preMove/postMove inside ObjectModel  we'll need to keep `OMR_VMThread` and `movedObjectHashCode` in OMR side which is the main reason we're doing this. 
- Move `movedObjectHashCode` to J9VMThread and keep preMove/postMove inside ObjectModel, which is slightly better than current implementation

bragaigor replied:
The current state of the PR reflects approach **[1]** from the above comment.

amicic replied:
We do have a pair of similar methods in EnvBase already: save/restoreObject, that deal with whole objects before/after GC in some special cases. We do similar things but with specific object flags before after object movement. So it's not totally out of place, although not perfect fit either.

I'm not totally convinced moving getting CompactSchemeFixupObject involved would help, although I have not put much thought. 

Anyhow, I'm ok with the current approach, but if you prefer [2] or [3] more, I'd probably be with them too.

bragaigor replied:
CompactSchemeFixupObject wouldn't help at all, I was brainstorming. And sounds good, that's true I did not consider save/restoreObject before. Okay then let's stick to this approach, I'll address your latest comments where I'll update preMove/postMove to the more appropriate names and add some comments there as well

amicic commented on runtime/gc_vlhgc/WriteOnceCompactor.cpp:
```diff
@@ -1182,6 +1167,35 @@ MM_WriteOnceCompactor::fixupObjectsInRange(MM_EnvironmentVLHGC *env, void *lowAd
+}
+
+void
+MM_WriteOnceCompactor::postObjectMove(MM_EnvironmentVLHGC* env, J9Object *newLocation, J9Object *objectPtr)
```
those 2 should be inlined, since they are only called once

amicic commented on runtime/gc_glue_java/ArrayletObjectModel.hpp:
```diff
@@ -923,6 +923,36 @@ class GC_ArrayletObjectModel : public GC_ArrayletObjectModelBase
+	 * @param arrayPtr      Pointer to the indexable object whose size is required
+	 */
+	MMINLINE void
+	setObjectDataAddr(omrobjectptr_t arrayPtr)
```
now that is lost addr parameter, it could be 'fixup; (again, not that I like it over 'update', but it's just already widespread)

amicic commented on runtime/gc_glue_java/EnvironmentDelegate.hpp:
```diff
@@ -28,9 +28,17 @@
+typedef struct GCmovedObjectHashCode {
+	uint32_t originalHashCode;
+	BOOLEAN hasBeenMoved;
+	BOOLEAN hasBeenHashed;
```
I'd use C++ style 'bool'.

I personally don't like using C style typedef struct either, but it's widespread already, so you can leave it.

amicic commented on runtime/gc_glue_java/EnvironmentDelegate.hpp:
```diff
@@ -159,6 +169,39 @@ class MM_EnvironmentDelegate
+	preMove(MM_GCExtensionsBase *extensions, omrobjectptr_t objectPtr)
+	{
+
+		bool hashed = extensions->objectModel.hasBeenHashed(objectPtr);
```
perhaps you can even cache ObjectModel (and not Ext as previously suggested)?

_language_vm that you need later you can get from already cached Env.

amicic commented on runtime/gc_glue_java/EnvironmentDelegate.hpp:
```diff
@@ -159,6 +170,40 @@ class MM_EnvironmentDelegate
 	void forceOutOfLineVMAccess();
 
+	MMINLINE void
+	preMove(omrobjectptr_t objectPtr)
```
I'd actually stick 'Object' in the name (here and in EnvBase too)
While the parameter somewhat implies that we move object, class name  (Environment) or instance name (env0 when we invoke it, does not. So, a naive thought would be that we move Env (or something inside Env).

amicic replied:
It would also be nice if the name is descriptive enough to say that we are using this only for _sliding_ (as opposed to _evacuating_ like copy-forward)  object movement. The former can partially damage the source object, in particular object header if sliding is less the object size (rare but possible), while the latter cannot (dest object is in dedicated survivor area). This is why we need to save those flags pre movement, since we need them later (for potential object growth).

Anyhow, what I want to say, we could stick 'Compact' in the name, too (here and EnvBase, but not in CompactScheme/WriteOnceCompactor - it's clear there).

Or if you fill it's not necessary to change the name, at least you should put a comment for methods what kind of object movement this is/is not used for.

amicic replied:
I'm wasting out time on something not that important, but still, I realized that compact in preObjectCompact would imply that object is being compacted (while in fact the heap is being compacted). preObjectSlide would be correct (we are indeed sliding object), but we don't use 'slide' that frequently in code/comments, and perhaps perObjectCompactMove would be ok (Compact would be noun as adjective to Move: we move object, but movement is compact style. Pick one, or come up with something better :))

Anyhow, I think we are ready to start breaking this into smaller, serialized PRs....

bragaigor replied:
That makes sense, hmm.. What about `preObjectMoveForCompact` and `postObjectMoveForCompact`?

bragaigor commented:
Renamed from pre/postObjectCompact to pre/postObjectMoveForCompact, went through the code once more, rebased and squashed. Ready for another round of review :)

bragaigor commented:
I just wanted to pin here (the comment got lost somewhere) that we decided to go forward with https://github.com/eclipse/openj9/pull/11890#discussion_r575252566 approach, which is the current state of the PR.

amicic commented:
Jenkins compile all jdk11

amicic commented:
Jenkins sanity test win,aix jdk11

amicic commented:
Jenkins test sanity win,aix jdk11

amicic commented:
Jenkins test sanity pLinux jdk8

amicic commented on runtime/gc_vlhgc/WriteOnceCompactor.cpp:
```diff
@@ -1182,6 +1167,36 @@ MM_WriteOnceCompactor::fixupObjectsInRange(MM_EnvironmentVLHGC *env, void *lowAd
+{
+	*objectSizeAfterMove = _extensions->objectModel.getConsumedSizeInBytesWithHeaderForMove(objectPtr);
+	_extensions->objectModel.preMove(env->getOmrVMThread(), objectPtr); // To be deprecated
+	//env->preObjectMoveForCompact(objectPtr); // Will substitute line above
```
seems like there is an action due here?

bragaigor replied:
I thought to leave these here since they'll be part of the 3rd part of changes. Or should I remove it?

amicic replied:
it's ok if this is to be addressed in followup changes

bragaigor replied:
Yes they are!

amicic commented on runtime/gc_vlhgc/WriteOnceCompactor.cpp:
```diff
@@ -1182,6 +1167,36 @@ MM_WriteOnceCompactor::fixupObjectsInRange(MM_EnvironmentVLHGC *env, void *lowAd
+	}
+
+	_extensions->objectModel.postMove(env->getOmrVMThread(), newLocation); // To be deprecated
+	//env->postObjectMoveForCompact(newLocation); // Will substitute line above
```
and here?
